### ISSUE TITLE
AbEnd when insert a page pointer

### ISSUE THREAD

**Issue by technicbasics** [trevorsandy/lpub3d#789]
**Summary**
Every time when insert a page pointer, LPub3D crash.
It does not matter which pointer is inserted (top, right, left, bottom)

LPub3D Version

    Incarnation: Windows portable in admin mode
    Affected build version: 2.4.8r157
    Known working build version: Tried some versions back, 2.4.7r0 works

Runtime Environment

    Operating system:
    Windows 11 Home Version 24H2
    Hardware architecture:
    Prozessor 11th Gen Intel(R) Core(TM) i9-11900H @ 2.50GHz 2.50 GHz
    Installierter RAM 16,0 GB (15,8 GB verwendbar)
    Systemtyp 64-Bit-Betriebssystem, x64-basierter Prozessor

**Steps to Reproduce**

 1. Open model file
 2. select a step
 3. right click and "insert Page Pointer"

**Workaround** (optional)
Insert the line with the page pointer syntax manually into the text editor.

[LPub3D_dmp_pdb.zip](https://github.com/user-attachments/files/17805585/LPub3D_dmp_pdb.zip)

**Comment by trevorsandy** [comment]
Thank you for reporting this behaviour. I'll take a look.

Cheers,

**Comment by trevorsandy** [comment]
I'm not able to reproduce the behaviour on the latest DevOps branch. 

Have a look at this video and let me know if there is any additional configuration needed.

https://github.com/user-attachments/assets/d78b4ae8-de92-4ee5-bc63-6d3dcafee612

Cheers,

**Comment by technicbasics** [comment]
Maybe my file is the problem.
I will test a other file asap.

Regards

20.11.2024 01:16:23 Trevor SANDY ***@***.***>:

**Comment by technicbasics** [comment]
Sorry, I read your reply sloppily. I first understood/read that you could not recreate the behavior.
Many, many thanks for your tireless work.

**Comment by trevorsandy** [comment]
> Sorry, I read your reply sloppily. I first understood/read that you could not recreate the behavior.
Many, many thanks for your tireless work.

No worries - I actually posted my initial findings a bit prematurely. Nevertheless, the latest v2.4.8 [DevOps](https://github.com/trevorsandy/lpub3d-ci/releases/tag/continuous) build is available.

If your checks are all good, this will be the last round of stabilization before production release.

Cheers,

**Comment by technicbasics** [comment]
Hi Trevor
I have now run what feels like 100 tests (in reality it was about 40-50), and I am almost desperate. Sometimes it doesn't work, then it works again. At first I thought it was a problem with the graphics card, as I force LPub3D to run via the NVIDIA card. I deactivated this and then it worked. Enabled this option again, it always works. Tried different files, same behavior. It doesn't work x times, then it works again x times etc....
With my first files I also had the problem that I could not save the file so quickly before the window closed. With a larger file I managed to save it after the error message before the window closed. And now I think I know at least when the program crashes and when it doesn't. However, I cannot say how the problem arises or in what context (sequence of code lines,...) it arises.
Long story short:
If the LPUB meta is at the beginning of the step, the program crashes...
![grafik](https://github.com/user-attachments/assets/d80dbd6f-5bd3-4398-8ffa-e5a570ba5282)

If the meta is at the end, the program runs...
![grafik](https://github.com/user-attachments/assets/a06cbc41-3eab-4852-9643-59824b5a81d4)

I suppose, depending on where the meta command is set when inserting, the program crashes or not.

Regards Johann

Translated with DeepL.com (free version)

**Comment by trevorsandy** [comment]
Yep - At the moment, the page pointer object needs a valid STEP object for the page it is assigned to. I'm looking at changing this - a page object should be enough.

When you manually place the `0 !LPUB PAGE POINTER...` command at the top of a `STEP`, no step object yet exists so '_boom!_' it'll blow up. At this moment, a `STEP` object is created when a valid `STEP` trigger is encountered - e.g. a part, an inserted page, as display submodel etc... So an alternative to removing the `STEP` dependency is to trigger one when a page pointer is encountered. 

Cheers,

**Comment by trevorsandy** [comment]
This behaviour has been changed.

The implemented design is to check if the command is at the top of the `STEP` and if we are, check if there are any parts or other step triggering item, including substitute parts, in the `STEP`. The `STEP` dependency, as discussed above, cannot be removed as we need to know where the step's assembly is to properly place the pointer tip so this design is not credible.

If there is a valid step item, e.g. a part, in the current `STEP`, then trigger the step object creation and proceed.
![Screenshot 2024-11-22 210804](https://github.com/user-attachments/assets/51447fe7-6718-4164-89b7-9a6a64079926)

If we do not have a valid `STEP`, one which has an assembly of some sort, then trigger a parse error message. 
![Screenshot 2024-11-22 212509](https://github.com/user-attachments/assets/1be5402e-97e9-4d7e-97e0-057ee4a88b2e)

If the message is ignored, then proceed without creating a pointer object - how about my aspirational cover page image :-)
![Screenshot 2024-11-22 212547](https://github.com/user-attachments/assets/1d0a24e0-fc72-4945-bdb7-9ff4cc31f826)

Cheers,

**Comment by technicbasics** [comment]
Hi Trevor
Page pointer in the r175 works perfectly. Every now and then I get an AbEnd after the export as png when the export is finished and the window to open the path where the export was done should be displayed. As I said it is sporadic, if I can discover any connection in the future I will of course report back with my findings. So far the version is running flawlessly. Thanks again for your work.

**Comment by trevorsandy** [comment]
Hi Johann,

Very well. If you are able to retrieve a .dmp file for your AbEnd that would be super also.

Cheers,

**Comment by technicbasics** [comment]
> Hi Johann,
> 
> Very well. If you are able to retrieve a .dmp file for your AbEnd that would be super also.
> 
> Cheers,

Hi Trevor
After a long time, here is a .dmp file after a AbEnd after exporting to png.
LPub3D version is r175.

[LPub3D.dmp](https://github.com/user-attachments/files/17962084/LPub3D.dmp)

**Comment by trevorsandy** [comment]
Thank you for this. I did not mention but should have that is is imperative to have the .pdb along with the .dmp. I will have to recompile r175 as the latest DevOps build is now at r201 and doing so will produce a .pdb with different signatures even though the code is the same. As a rule of thumb, it is a good practice to provide the .exe, .pdb, and .dmp together if you can.

Cheers,

**Comment by technicbasics** [comment]
No problem..
[LPub3D_exe.zip](https://github.com/user-attachments/files/17962600/LPub3D_exe.zip)
[LPub3D_pdb.zip](https://github.com/user-attachments/files/17962601/LPub3D_pdb.zip)
